src/v/cluster/scheduling/leader_balancer.cc

```diff
@@ -0,0 +1,519 @@
+/*
+ * Copyright 2020 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "cluster/scheduling/leader_balancer.h"
+
+#include "cluster/logger.h"
+#include "cluster/partition_leaders_table.h"
+#include "cluster/scheduling/leader_balancer_greedy.h"
+#include "cluster/shard_table.h"
+#include "cluster/topic_table.h"
+#include "model/namespace.h"
+#include "rpc/types.h"
+#include "seastarx.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/core/timer.hh>
+
+#include <fmt/core.h>
+#include <fmt/format.h>
+#include <fmt/ostream.h>
+#include <fmt/ranges.h>
+
+#include <chrono>
+
+namespace cluster {
+
+leader_balancer::leader_balancer(
+  topic_table& topics,
+  partition_leaders_table& leaders,
+  raft::consensus_client_protocol client,
+  ss::sharded<shard_table>& shard_table,
+  ss::sharded<partition_manager>& partition_manager,
+  ss::sharded<raft::group_manager>& group_manager,
+  ss::sharded<ss::abort_source>& as,
+  consensus_ptr raft0)
+  : _topics(topics)
+  , _leaders(leaders)
+  , _client(std::move(client))
+  , _shard_table(shard_table)
+  , _partition_manager(partition_manager)
+  , _group_manager(group_manager)
+  , _as(as)
+  , _raft0(std::move(raft0))
+  , _timer([this] { trigger_balance(); }) {}
+
+ss::future<> leader_balancer::start() {
```

Comment:
Would be good to have a top level configuration option to enable/disable the rebalancing in case of issues.  Could combine it with a setting to configure the idle timeout

Reply:
Agreed!